File: Benchmarks_FSharp/Core/Scene/EventManager.fs
namespace Benchmarks.Core.Scene

open System
open System.Collections.Generic

type EventManager() =
    let _subscribers = Dictionary<Type, List<Delegate>>()

    static let mutable _instance: EventManager = EventManager()

    static member Instance
        with get() = _instance

    member _.Subscribe<'T>(handler: Action<'T>) =
        let t = typeof<'T>
        let mutable list = Unchecked.defaultof<List<Delegate>>
        if not (_subscribers.TryGetValue(t, &list)) then
            list <- List<Delegate>()
            _subscribers.[t] <- list
        list.Add(handler)

    member _.Publish<'T>(eventData: 'T) =
        let t = typeof<'T>
        match _subscribers.TryGetValue(t) with
        | true, list ->
            for handler in list do
                (handler :?> Action<'T>).Invoke(eventData)
        | _ -> ()

    member _.Clear() =
        _subscribers.Clear()

    static member Reset() =
        _instance <- EventManager()

File: Benchmarks/Core/Scene/EventManager.cs
namespace Benchmarks.Core.Scene
{
    /// <summary>
    /// Simple pub/sub event manager that replicates the engine's EventManager pattern.
    /// </summary>
    public class EventManager
    {
        private static EventManager _instance;
        public static EventManager Instance => _instance ??= new EventManager();

        private readonly Dictionary<Type, List<Delegate>> _subscribers = new();

        public void Subscribe<T>(Action<T> handler)
        {
            var type = typeof(T);
            if (!_subscribers.TryGetValue(type, out var list))
            {
                list = new List<Delegate>();
                _subscribers[type] = list;
            }
            list.Add(handler);
        }

        public void Publish<T>(T eventData)
        {
            var type = typeof(T);
            if (_subscribers.TryGetValue(type, out var list))
            {
                foreach (var handler in list)
                {
                    ((Action<T>)handler)(eventData);
                }
            }
        }

        public void Clear()
        {
            _subscribers.Clear();
        }

        public static void Reset()
        {
            _instance = new EventManager();
        }
    }
}
